File: Data Structures & Algorithms/house-robber-ii/submission-1.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) {
            return nums[0];
        }
        int a = 0, b = nums[0];
        for (int i = 1; i < n-1; i++) {
            int pick = nums[i] + a;
            int not_pick = b;
            a = b;
            b = max(pick, not_pick);
        }
        int res = b;
        a = 0, b = nums[1];
        for (int i = 2; i < n; i++) {
            int pick = nums[i] + a;
            int not_pick = b;
            a = b;
            b = max(pick, not_pick);
        }
        res = max(res, b);
        return res;
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [-1]*n

        def dfs(i):
            if dp[i] != -1:
                return dp[i]

            LIS = 1
            for j in range (i + 1, n):
                if nums[i] < nums[j]:
                    LIS = max(LIS, 1 + dfs(j))

            dp[i] = LIS
            return LIS

        return max(dfs(i) for i in range(n))
        